src/reconcile/rec_visibility.c

```diff
@@ -580,6 +549,200 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, W
             break;
     }
 
+    /*
+     * Track the most recent transaction in the page. We store this in the tree at the end of
+     * reconciliation in the service of checkpoints, it is used to avoid discarding trees from
+     * memory when they have changes required to satisfy a snapshot read.
+     */
+    if (WT_TXNID_LT(r->max_txn, max_txn))
+        r->max_txn = max_txn;
+
+    /* Update the maximum timestamp. */
+    if (max_ts > r->max_ts)
+        r->max_ts = max_ts;
+
+    /*
+     * We should never select an update that has been written to the history store except checkpoint
+     * writes the update that is older than a prepared update or we need to first delete the update
+     * from the history store.
+     */
+    WT_ASSERT(session,
+      upd_select->upd == NULL || !F_ISSET(upd_select->upd, WT_UPDATE_HS) ||
+        F_ISSET(upd_select->upd, WT_UPDATE_TO_DELETE_FROM_HS) ||
+        (!F_ISSET(r, WT_REC_EVICT) && seen_prepare));
+
+    return (0);
+}
+
+/*
+ * __rec_fill_upd_select --
+ *     Fill the time window information and the selected update.
+ */
+static int
+__rec_fill_upd_select(
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_CELL_UNPACK_KV *vpack, WT_UPDATE_SELECT *upd_select)
+{
+    WT_TIME_WINDOW *select_tw;
+    WT_UPDATE *last_upd, *upd, *tombstone;
+
+    upd = upd_select->upd;
+    last_upd = tombstone = NULL;
+    select_tw = &upd_select->tw;
+
+    /*
+     * The start timestamp is determined by the commit timestamp when the key is first inserted (or
+     * last updated). The end timestamp is set when a key/value pair becomes invalid, either because
+     * of a remove or a modify/update operation on the same key.
+     */
+
+    /*
+     * Mark the prepare flag if the selected update is an uncommitted prepare. As tombstone updates
+     * are never returned to write, set this flag before we move into the previous update to write.
+     */
+    if (upd->prepare_state == WT_PREPARE_INPROGRESS)
+        select_tw->prepare = 1;
+
+    /*
+     * If the newest is a tombstone then select the update before it and set the end of the
+     * visibility window to its time point as appropriate to indicate that we should return "not
+     * found" for reads after this point.
+     *
+     * Otherwise, leave the end of the visibility window at the maximum possible value to indicate
+     * that the value is visible to any timestamp/transaction id ahead of it.
+     */
+    if (upd->type == WT_UPDATE_TOMBSTONE) {
+        WT_TIME_WINDOW_SET_STOP(select_tw, upd);
+        tombstone = upd;
+        upd_select->tombstone = upd;
```

Comment:
Is there any difference between `tombstone` and `upd_select->tombstone` here?
It looks like it'd be cleaner to set `tombstone = upd_select->tombstone` and only refer to `tombstone` in the function.